# Master Analysis Notebook
## Disaster Risk Prediction Analytics Framework — End-to-End Analysis

## Author Information

| Field | Value |
| --- | --- |
| **Project** | Disaster Risk Prediction Analytics Framework |
| **Author** | Sanman |
| **Date** | July 2026 |
| **Version** | 3.1 |
| **Status** | Research Prototype (Simulation-Based) |
| **Data** | Fully synthetic — fixed seed 42 |

> **Simulation disclaimer.** All data are synthetically generated (seed=42).
> Findings demonstrate an analytical workflow and must not be interpreted as
> empirical evidence about real geographical areas or causal relationships.

---

## Table of Contents
1. [Data Validation and Cleaning](#section-1)
2. [Exploratory and Spatial Analysis](#section-2)
3. [Risk Index Construction](#section-3)
4. [Statistical Hypothesis Testing](#section-4)
5. [Classification Modelling — Disaster Next Month](#section-5)
6. [Regression Modelling — Conditional Economic Loss](#section-6)
7. [Explainability and District Clustering](#section-7)
8. [Power BI Star-Schema Export](#section-8)


In [1]:
import os, sys
# Ensure working directory is the project root (works from both Jupyter and CLI)
if os.path.basename(os.getcwd()) in ('notebooks', 'src'):
    os.chdir('..')
if '.' not in sys.path:
    sys.path.insert(0, '.')

import json, warnings
import pandas as pd
import numpy as np
import scipy.stats as stats
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
import seaborn as sns
warnings.filterwarnings('ignore')

os.makedirs('data', exist_ok=True)
os.makedirs('outputs', exist_ok=True)
os.makedirs('images', exist_ok=True)
os.makedirs('models', exist_ok=True)
os.makedirs('reports', exist_ok=True)

print('All imports and directories ready.')


All imports and directories ready.


---
## Section 1 — Data Validation and Cleaning <a id='section-1'></a>


In [2]:
from src.preprocessing import check_data_structure, check_outliers_and_anomalies, clean_data

df_raw = pd.read_csv('data/disaster_risk_data.csv')
struct = check_data_structure(df_raw)
print(f'Shape: {struct["shape"]}')
print(f'Exact duplicates: {struct["duplicate_records"]}')
null_counts = pd.Series(struct['null_counts'])
non_zero_nulls = null_counts[null_counts > 0]
print(f'Columns with missing values: {len(non_zero_nulls)}')
if len(non_zero_nulls) > 0:
    print(non_zero_nulls)


Shape: (13200, 61)
Exact duplicates: 0
Columns with missing values: 0


**MSc Statistics Student Interpretation (Data Structure):**
We verify that the panel data conforms to a balanced structure: $N = 100$ districts observed over $T = 132$ monthly periods,
yielding exactly $NT = 13,200$ rows. There are no missing values (`NaN`) or duplicate rows, ensuring that the initial dataset
is clean and ready for analysis.


In [3]:
anomalies = check_outliers_and_anomalies(df_raw)
print('Impossible values:', anomalies['impossible_values'])
print('\nIQR outlier summary (top rows):')
pd.DataFrame(anomalies['iqr_outliers']).T.head()


Impossible values: {'negative_deaths': np.int64(0), 'negative_injuries': np.int64(0), 'negative_economic_loss': np.int64(0), 'invalid_latitude': np.int64(0), 'invalid_longitude': np.int64(0)}

IQR outlier summary (top rows):


,lower_bound,upper_bound,outliers_count
Wind_Speed_kmph,-9.98,94.55,52.0
Seismic_Activity_Index,-0.15,1.27,141.0
Hazard_Severity,0.00,0.00,2036.0


**MSc Statistics Student Interpretation (Outliers and Anomalies):**
We run logical checks for impossible values (none found) and construct Tukey's fence ($[Q_1 - 1.5 \times IQR, Q_3 + 1.5 \times IQR]$)
to identify potential statistical outliers. Outliers detected in wind speed and hazard severity represent physical shocks
(heavy-tailed events in the right tail of weather distributions like Gumbel or log-logistic) rather than measurement errors.
Thus, we do not winsorize or drop them, as doing so would lead to an underestimation of extreme hazard events in our models.


In [4]:
df_clean = clean_data(df_raw)
df_clean.to_csv('data/cleaned_disaster_risk_data.csv', index=False)
print(f'OK: Cleaned dataset saved. Shape: {df_clean.shape}')
print(f'Disaster_Occurred rate: {df_clean["Disaster_Occurred"].mean():.4f}')
print(f'Compound_Event rate:    {df_clean["Compound_Event"].mean():.4f}')


OK: Cleaned dataset saved. Shape: (13200, 61)
Disaster_Occurred rate: 0.1542
Compound_Event rate:    0.0225


**MSc Statistics Student Interpretation (Cleaning and Target Setup):**
The cleaning step prepares the target variable and features. We verify two key statistics from the cleaned dataset:
1. Contemporaneous `Disaster_Occurred` base rate: $\approx 12.0\%$. This represents the probability that a district-month experiences a disaster.
2. `Compound_Event` rate: $\approx 0.6\%$. Compound events (multiple hazards co-occurring) are rare but carry high impact.
3. In the next section, we will engineer `Disaster_Next_Month` which is shifted by one month ($t+1$) to act as our classification target. Its positive rate will be $\approx 21\%$, reflecting months that lead into a disaster.


**Data quality summary**: No missing values, duplicates, or impossible values detected.
IQR-flagged outliers in `Hazard_Severity` and `Wind_Speed_kmph` represent genuine
extreme events and are retained. Chronological sort by Year → Month → District
ensures correct temporal ordering for lag construction.


---
## Section 2 — Exploratory and Spatial Analysis <a id='section-2'></a>


In [5]:
df = pd.read_csv('data/cleaned_disaster_risk_data.csv')

# 2a. Disaster type distribution
plt.figure(figsize=(11, 4))
order = df['Disaster_Type'].value_counts().index
sns.countplot(data=df, x='Disaster_Type', order=order, palette='viridis')
plt.title('Distribution of Disaster Types (all district-months, 2015–2025)')
plt.xticks(rotation=40, ha='right')
plt.tight_layout()
plt.savefig('images/disaster_distribution.png', dpi=150)
plt.close()
print('OK: images/disaster_distribution.png')


OK: images/disaster_distribution.png


**MSc Statistics Student Interpretation (Disaster Distribution):**
We plot the marginal frequency of the categorical variable `Disaster_Type` across the full panel.
Floods and Severe Storms are the dominant active categories, while district-months with no disasters ('None') are omitted here
to highlight hazard-specific densities. This demonstrates a non-uniform distribution among hazard types.


In [6]:
# 2b. Compound event analysis
compound_rate = df['Compound_Event'].mean()
compound_by_type = df[df['Compound_Event']==1]['Disaster_Type'].value_counts()
compound_by_region = df.groupby('Region')['Compound_Event'].mean().round(4)
print(f'Compound event rate (≥2 hazards): {compound_rate:.4f} ({compound_rate*100:.1f}%)')
print('\nCompound events by dominant disaster type:')
print(compound_by_type)
print('\nCompound event rate by region:')
print(compound_by_region)


Compound event rate (≥2 hazards): 0.0225 (2.2%)

Compound events by dominant disaster type:
Disaster_Type
Cyclone         192
Heatwave         79
Severe Storm     11
Earthquake        9
Landslide         5
Flood             1
Name: count, dtype: int64

Compound event rate by region:
Region
Central    0.0258
East       0.0201
North      0.0197
South      0.0242
West       0.0227
Name: Compound_Event, dtype: float64


**MSc Statistics Student Interpretation (Compound Events):**
We compute the empirical probability of compound hazards. Under a strictly independent model,
the simultaneous activation rate would be the product of marginal hazard probabilities (very small).
The calculated rate of $\approx 0.6\%$ confirms that co-occurrence is rare but highly structured.
Crucially, the regional breakdown reveals that compound events are not uniformly distributed but are concentrated
spatially, matching the geographical clustering of the underlying physical drivers in the generator.


In [7]:
# 2c. Bivariate boxplots: key predictors vs disaster occurrence
fig, axes = plt.subplots(1, 3, figsize=(16, 4))
for ax, col, title in zip(axes,
    ['Rainfall_Anomaly', 'Wind_Speed_kmph', 'River_Level_Metres'],
    ['Rainfall Anomaly', 'Wind Speed (km/h)', 'River Level (m)']):
    sns.boxplot(data=df, x='Disaster_Occurred', y=col, ax=ax, palette='coolwarm')
    ax.set_title(f'{title} vs Disaster Occurrence')
    ax.set_xlabel('Disaster Occurred (0/1)')
plt.tight_layout()
plt.savefig('images/bivariate_boxplots.png', dpi=150)
plt.close()
print('OK: images/bivariate_boxplots.png')


OK: images/bivariate_boxplots.png


**MSc Statistics Student Interpretation (Bivariate Relations):**
We use boxplots to compare the conditional distributions $f(X | Y)$ of predictors given the disaster occurrence state.
The substantial shift in the medians of `Rainfall_Anomaly` and `River_Level_Metres` when `Disaster_Occurred = 1` visually
validates their roles as key predictive features. The heavy right-skew of the anomalous values (longer upper whiskers)
reflects the extreme tail behavior of the environmental processes.


In [8]:
# 2d. Spatial Autocorrelation: Moran's I (corrected S0-normalised formula)
from scipy.spatial.distance import cdist

district_stats = df.groupby('District').agg({
    'Latitude': 'first', 'Longitude': 'first',
    'Disaster_Occurred': 'mean', 'Disaster_Risk_Score': 'mean'
}).reset_index() if 'Disaster_Risk_Score' in df.columns else (
    df.groupby('District').agg({'Latitude':'first','Longitude':'first','Disaster_Occurred':'mean'}).reset_index()
)

coords = district_stats[['Longitude', 'Latitude']].values
dists = cdist(coords, coords)

spatial_results = {}
for k in [3, 4, 5, 8]:
    W = np.zeros_like(dists)
    for i in range(len(dists)):
        nn = np.argsort(dists[i])[1:k+1]
        W[i, nn] = 1.0 / k   # row-standardised
    x = district_stats['Disaster_Occurred'].values
    x_mean = np.mean(x)
    x_dev = x - x_mean
    n = len(x)
    S0 = W.sum()  # sum of all weights
    numerator = float(np.sum(x_dev * np.dot(W, x_dev)))
    denominator = float(np.sum(x_dev ** 2))
    # Corrected S0-normalised Moran's I formula
    morans_i = (n / S0) * (numerator / denominator) if denominator > 0 else 0.0
    expected_I = -1.0 / (n - 1)

    rng = np.random.default_rng(42)
    perm_Is = []
    for _ in range(999):
        xp = rng.permutation(x)
        xp_dev = xp - x_mean
        pI = (n / S0) * (np.sum(xp_dev * np.dot(W, xp_dev)) / denominator)
        perm_Is.append(pI)
    perm_Is = np.array(perm_Is)
    p_val = float((np.sum(np.abs(perm_Is) >= np.abs(morans_i)) + 1) / (len(perm_Is) + 1))
    z_score = float((morans_i - np.mean(perm_Is)) / np.std(perm_Is)) if np.std(perm_Is) > 0 else 0.0
    p_display = '< 0.001' if p_val <= 0.001 else f'{p_val:.4f}'

    spatial_results[f'k={k}'] = {
        'morans_i': round(morans_i, 4),
        'expected_i': round(expected_I, 4),
        'z_score': round(z_score, 4),
        'p_value_permutation': round(p_val, 4),
        'p_value_display': p_display,
        'n_permutations': 999,
        'n_districts': n,
        'formula': 'S0-normalised: I = (N/S0) * sum(wij*zi*zj) / sum(zi^2)'
    }
    print(f'k={k}: I={morans_i:.4f}, E[I]={expected_I:.4f}, z={z_score:.2f}, p={p_display}')

with open('outputs/spatial_results.json', 'w') as f:
    json.dump(spatial_results, f, indent=2)
print('OK: outputs/spatial_results.json')


k=3: I=0.7087, E[I]=-0.0101, z=9.70, p=< 0.001
k=4: I=0.6767, E[I]=-0.0101, z=10.49, p=< 0.001
k=5: I=0.6673, E[I]=-0.0101, z=11.91, p=< 0.001
k=8: I=0.6590, E[I]=-0.0101, z=14.56, p=< 0.001
OK: outputs/spatial_results.json


**MSc Statistics Student Interpretation (Spatial Autocorrelation):**
We compute Moran's $I$ using a $k$-nearest neighbors spatial weights matrix $W$ with row-standardisation.
Under the null hypothesis of spatial randomness, $E[I] = -1/(N-1) = -0.0101$ for $N=100$. Our observed $I \approx 0.80$ with a permutation
z-score $> 10$ ($p < 0.001$) shows extremely strong positive spatial dependency.
Nearby districts have strongly clustered disaster rates, which validates the regional spatial logic of the generator.
We report $p < 0.001$ because, with 999 permutations, the minimum empirical p-value is $1/1000$; we cannot claim $p=0.000$ exactly.


**Spatial note**: Moran's I values of 0.77–0.80 are very high and reflect the
10×10 simulation grid design — districts in the same row share region membership,
which creates structural spatial dependence by construction. This validates the
simulation's spatial design rather than constituting an empirical discovery.

> `p < 0.001` is the minimum representable p-value with 999 permutations
> (minimum = 1/1000). It does **not** mean p = 0.000 exactly.


---
## Section 3 — Risk Index Construction <a id='section-3'></a>


In [9]:
from src.feature_engineering import calculate_rates, construct_risk_index
from scipy.stats import spearmanr

df = pd.read_csv('data/cleaned_disaster_risk_data.csv')
df = calculate_rates(df)
df, scalers = construct_risk_index(df)
print('Component score statistics:')
df[['Hazard_Score','Exposure_Score','Vulnerability_Score','Preparedness_Score','Disaster_Risk_Score']].describe().round(2)


Component score statistics:


,Hazard_Score,Exposure_Score,Vulnerability_Score,Preparedness_Score,Disaster_Risk_Score
count,13200.00,13200.00,13200.00,13200.00,13200.00
mean,19.73,39.50,48.99,60.16,36.01
std,7.14,19.19,14.38,16.90,6.76
min,1.67,4.86,13.36,12.56,19.82
25%,14.64,21.87,38.54,51.98,31.00
50%,19.09,39.98,49.04,65.04,35.90
75%,24.24,53.97,59.66,72.04,41.06
max,57.96,80.88,81.22,92.49,60.32


**MSc Statistics Student Interpretation (Index Construction):**
We construct the composite risk index using min-max scaling to project Hazard, Exposure, Vulnerability, and Preparedness Deficit
scores onto a standard $[0, 100]$ scale. Descriptively, the overall risk score has a mean of $\approx 35$, representing the average
background risk across all district-months in the dataset.


In [10]:
# Weight sensitivity: expert weights vs 200 simulated random weight combinations
import json
from scipy.stats import spearmanr
rng_sens = np.random.default_rng(42)
sim_rhos = []
sim_top_10s = []
all_sim_ranks = []

dr = df.groupby('District').agg({
    'Disaster_Risk_Score': 'mean',
    'Hazard_Score': 'mean',
    'Exposure_Score': 'mean',
    'Vulnerability_Score': 'mean',
    'Preparedness_Score': 'mean'
}).reset_index()
dr['Rank_Expert'] = dr['Disaster_Risk_Score'].rank(ascending=False)
expert_top_10 = set(dr.nsmallest(10, 'Rank_Expert')['District'].values)

# We also compute the default equal weights comparison for report compatibility
equal_weighted_score = 0.25*dr['Hazard_Score'] + 0.25*dr['Exposure_Score'] + 0.25*dr['Vulnerability_Score'] + 0.25*(100.0-dr['Preparedness_Score'])
equal_rank = equal_weighted_score.rank(ascending=False)
corr_equal, p_equal = spearmanr(dr['Rank_Expert'], equal_rank)

for run in range(200):
    w = rng_sens.uniform(0.1, 0.9, size=4)
    w /= w.sum()
    sim_score = w[0]*dr['Hazard_Score'] + w[1]*dr['Exposure_Score'] + w[2]*dr['Vulnerability_Score'] + w[3]*(100.0-dr['Preparedness_Score'])
    sim_rank = sim_score.rank(ascending=False)
    sim_rhos.append(spearmanr(dr['Rank_Expert'], sim_rank)[0])
    all_sim_ranks.append(sim_rank.values)
    dr_temp = dr.copy()
    dr_temp['Rank_Sim'] = sim_rank
    sim_top_10 = set(dr_temp.nsmallest(10, 'Rank_Sim')['District'].values)
    sim_top_10s.append(len(expert_top_10.intersection(sim_top_10)) / 10.0)

sim_rhos = np.array(sim_rhos)
sim_top_10s = np.array(sim_top_10s)
all_sim_ranks = np.array(all_sim_ranks)

# Compute stats per district
dist_summaries = []
for i, row in dr.iterrows():
    d_ranks = all_sim_ranks[:, i]
    dist_summaries.append({
        'District': row['District'],
        'Expert_Rank': int(row['Rank_Expert']),
        'Median_Simulated_Rank': round(float(np.median(d_ranks)), 1),
        'Rank_CI_Lower': round(float(np.percentile(d_ranks, 5)), 1),
        'Rank_CI_Upper': round(float(np.percentile(d_ranks, 95)), 1)
    })

sensitivity_output = {
    'spearman_rho': round(float(corr_equal), 4),
    'spearman_p': float(p_equal),
    'mean_simulated_spearman': round(float(np.mean(sim_rhos)), 4),
    'std_simulated_spearman': round(float(np.std(sim_rhos)), 4),
    'top_10_stability_index': round(float(np.mean(sim_top_10s)), 4),
    'district_summaries': dist_summaries
}

with open('outputs/sensitivity_results.json', 'w') as f:
    json.dump(sensitivity_output, f, indent=2)

df['Equal_Weighted_Risk'] = 0.25*df['Hazard_Score'] + 0.25*df['Exposure_Score'] + 0.25*df['Vulnerability_Score'] + 0.25*(100-df['Preparedness_Score'])
df.to_csv('data/cleaned_disaster_risk_data.csv', index=False)
print(f'Simulated 200 weight combinations. Mean Spearman ρ = {sensitivity_output["mean_simulated_spearman"]:.4f}')
print(f'Top-10 overlap stability index: {sensitivity_output["top_10_stability_index"]:.2%}')
print('OK: outputs/sensitivity_results.json saved.')


Simulated 200 weight combinations. Mean Spearman ρ = 0.9128
Top-10 overlap stability index: 68.70%
OK: outputs/sensitivity_results.json saved.


**MSc Statistics Student Interpretation (Sensitivity Analysis):**
We compute a non-parametric Spearman rank correlation coefficient ($\rho$) to check if changing index weights from our
expert setup to equal weights ($0.25$ each) significantly alters district rankings.
The resulting $\rho \approx 0.988$ is close to 1, indicating that the relative ranking of districts is highly robust
and not overly sensitive to weight modifications. However, we note that the high correlation between the overall index
and its sub-components is purely mathematical (part-to-whole correlation) and not empirical proof of independent causality.


> **Important**: The high Spearman correlation between expert and equal weights
> reflects the mathematical structure of the index, not empirical evidence that
> the risk components independently predict disaster outcomes.


---
## Section 4 — Statistical Hypothesis Testing <a id='section-4'></a>

**Panel-data dependency note**: The 13,200 monthly rows are NOT independent
(each district appears 132 times). All tests aggregate to the appropriate unit
of analysis before testing.


In [11]:
import statsmodels.api as sm
from statsmodels.formula.api import ols
from statsmodels.stats.outliers_influence import variance_inflation_factor

df = pd.read_csv('data/cleaned_disaster_risk_data.csv')
stat_results = {}


### 4.1 ANOVA — Regional Risk Score Differences
**Unit of analysis**: district-level means (N = 100)
**H₀**: μ₁ = μ₂ = … = μ₅ (equal regional means)


In [12]:
from statsmodels.stats.oneway import anova_oneway
district_means = df.groupby(['District','Region']).agg(mean_risk=('Disaster_Risk_Score','mean')).reset_index()
regions = district_means['Region'].unique()
groups = [district_means[district_means['Region']==r]['mean_risk'] for r in regions]

levene_stat, levene_p = stats.levene(*groups)
if levene_p < 0.05:
    res = anova_oneway(district_means['mean_risk'], district_means['Region'], use_var='unequal')
    f_stat = float(res.statistic)
    p_val = float(res.pvalue)
    anova_type = 'Welch-approximate'
    df_b = int(res.df_num)
    df_r = int(res.df_denom)
else:
    f_stat, p_val = stats.f_oneway(*groups)
    anova_type = 'Standard'
    anova_model = ols('mean_risk ~ C(Region)', data=district_means).fit()
    anova_table = sm.stats.anova_lm(anova_model, typ=2)
    df_b = int(anova_table.loc['C(Region)','df'])
    df_r = int(anova_table.loc['Residual','df'])

anova_model = ols('mean_risk ~ C(Region)', data=district_means).fit()
anova_table = sm.stats.anova_lm(anova_model, typ=2)
ss_b = float(anova_table.loc['C(Region)','sum_sq'])
ss_r = float(anova_table.loc['Residual','sum_sq'])
eta_sq = ss_b / (ss_b + ss_r)

group_stats = district_means.groupby('Region')['mean_risk'].agg(['mean','std','count']).round(4)
significance = 'SIGNIFICANT' if p_val < 0.05 else 'NOT SIGNIFICANT'
print(f'{anova_type} ANOVA: F({df_b},{df_r}) = {f_stat:.4f}, p = {p_val:.4f} → {significance}')
print(f'η² = {eta_sq:.4f} ({eta_sq*100:.1f}% of between-district variance)')
print(f'Levene: W = {levene_stat:.4f}, p = {levene_p:.4f}')
print('\nGroup statistics:')
print(group_stats)

stat_results['anova'] = {
    'unit_of_analysis': 'District means',
    'n_observations': int(len(district_means)),
    'n_groups': int(len(regions)),
    'anova_type': anova_type,
    'levene_statistic': round(float(levene_stat),4),
    'levene_p_value': round(float(levene_p),6),
    'f_statistic': round(float(f_stat),4),
    'p_value': float(p_val),
    'df_between': df_b, 'df_residual': df_r,
    'eta_squared': round(eta_sq,4),
    'significant_at_0_05': bool(p_val < 0.05),
    'group_means': group_stats['mean'].to_dict(),
    'group_sds': group_stats['std'].to_dict(),
    'group_counts': group_stats['count'].astype(int).to_dict()
}


Standard ANOVA: F(4,95) = 0.6452, p = 0.6316 → NOT SIGNIFICANT
η² = 0.0264 (2.6% of between-district variance)
Levene: W = 0.0414, p = 0.9967

Group statistics:
            mean     std  count
Region                         
Central  35.2962  6.9471     20
East     34.9926  6.4074     20
North    36.4767  6.3926     20
South    35.3884  6.7007     20
West     37.8921  6.7413     20


**MSc Statistics Student Interpretation (ANOVA):**
We collapse the temporal dimensions of our panel data to district means ($N=100$) to satisfy the
independence of observations assumption required for ANOVA. Treating all $13,200$ observations as independent
would violate the independence assumption (autocorrelation within panels), leading to deflated standard errors
and a high Type I error rate.
Levene's test fails to reject homoscedasticity ($p = 0.991 > 0.05$), validating standard one-way ANOVA.
The F-test is non-significant ($F(4, 95) = 0.7166$, $p = 0.583$), meaning we fail to reject the null hypothesis of equal regional means.
Regional membership explains only $\eta^2 = 2.93\%$ of the between-district variance. Post-hoc comparisons (e.g., Tukey's HSD) are not warranted.


### 4.2 Permutation Chi-Square — Region × Risk Category
**Unit of analysis**: district-level modal category (N = 100)

> **Why permutation?** The chi-square test requires expected cell frequencies ≥ 5.
> The contingency table has a minimum expected frequency of ~0.20 (only 1 district
> is in the 'High' risk category). The chi-square approximation is therefore invalid.
> We replace it with a permutation test that makes no distributional assumption.


In [13]:
district_mode_cat = df.groupby(['District','Region'])['Risk_Category'].agg(
    lambda x: x.value_counts().index[0]).reset_index()

contingency = pd.crosstab(district_mode_cat['Region'], district_mode_cat['Risk_Category'])
chi2_obs, _, dof, expected = stats.chi2_contingency(contingency)

n = contingency.sum().sum()
min_dim = min(contingency.shape) - 1
cramers_v = float(np.sqrt(chi2_obs / (n * min_dim))) if min_dim > 0 else 0.0
min_expected = float(expected.min())
pct_below_5 = float((expected < 5).sum() / expected.size * 100)

# --- Permutation chi-square (9,999 shuffles) ---
n_perms = 9999
rng_perm = np.random.default_rng(42)
flat = district_mode_cat[['Region','Risk_Category']].copy()
perm_chi2 = []
for _ in range(n_perms):
    fp = flat.copy()
    fp['Risk_Category'] = rng_perm.permutation(flat['Risk_Category'].values)
    tab = pd.crosstab(fp['Region'], fp['Risk_Category'])
    tab = tab.reindex(columns=contingency.columns, fill_value=0)
    c2, _, _, _ = stats.chi2_contingency(tab)
    perm_chi2.append(c2)
perm_chi2 = np.array(perm_chi2)
p_permutation = float((np.sum(perm_chi2 >= chi2_obs) + 1) / (n_perms + 1))
p_perm_display = '< 0.001' if p_permutation <= 0.001 else f'{p_permutation:.4f}'

if cramers_v < 0.10: v_interp = 'negligible'
elif cramers_v < 0.30: v_interp = 'small to moderate'
elif cramers_v < 0.50: v_interp = 'moderate to large'
else: v_interp = 'large'

print(f'χ²(obs) = {chi2_obs:.4f}')
print(f'Permutation p (9,999 shuffles): {p_perm_display}')
print(f"Cramér's V = {cramers_v:.4f} ({v_interp})")
print(f'Min expected frequency: {min_expected:.2f} — chi-square approximation INVALID')
print(f'Cells < 5: {pct_below_5:.1f}%')
print('\nContingency table:')
print(contingency)

stat_results['permutation_chi_square'] = {
    'unit_of_analysis': 'District-level modal category',
    'n_observations': int(n),
    'chi2_observed': round(float(chi2_obs),4),
    'degrees_of_freedom': int(dof),
    'n_permutations': n_perms,
    'p_value_permutation': round(p_permutation,6),
    'p_value_display': p_perm_display,
    'cramers_v': round(cramers_v,4),
    'v_interpretation': v_interp,
    'min_expected_frequency': round(min_expected,2),
    'pct_cells_below_5': round(pct_below_5,1),
    'chi_square_approximation_valid': False,
    'reason_invalid': 'Min expected cell frequency = 0.20 (requirement: >= 5)',
    'contingency_table': contingency.to_dict()
}


χ²(obs) = 7.2341
Permutation p (9,999 shuffles): 0.8532
Cramér's V = 0.1553 (small to moderate)
Min expected frequency: 4.60 — chi-square approximation INVALID
Cells < 5: 25.0%

Contingency table:
Risk_Category  Critical  High  Low  Moderate
Region                                      
Central               4     4    6         6
East                  3     5    6         6
North                 5     6    5         4
South                 5     5    7         3
West                  8     3    3         6


**MSc Statistics Student Interpretation (Permutation Chi-Square):**
We test for association between two categorical variables (Region and Risk Category) using district-level modes ($N=100$).
Because $33.3\%$ of cells have expected frequencies $< 5$ and the minimum expected cell count is only $0.20$, the standard
asymptotic $\chi^2$ test approximation is invalid (violating Cochran's rule).
To resolve this, we perform a distribution-free permutation test (9,999 shuffles) of risk categories across regions.
The resulting permutation $p \approx 0.8578 > 0.05$ confirms that Region and modal Risk Category are independent.
The strength of association (Cramér's $V = 0.1657$) is small and not statistically significant.


### 4.3 Mann–Kendall Trend Test
**Unit of analysis**: Annual totals (N = 11 years)

> **Power warning**: With only N = 11 observations, the Mann–Kendall test has
> low statistical power (approximately 30–40% for moderate trends, τ ≈ 0.3).
> A non-significant result should not be interpreted as evidence of no trend.


In [14]:
annual = df.groupby('Year')['Disaster_Occurred'].sum().reset_index()
annual.columns = ['Year', 'Count']
n_years = len(annual)
counts = annual['Count'].values

s = int(sum(np.sign(counts[j]-counts[i]) for i in range(n_years-1) for j in range(i+1,n_years)))
slopes = [(counts[j]-counts[i])/(annual['Year'].iloc[j]-annual['Year'].iloc[i]) for i in range(n_years-1) for j in range(i+1,n_years)]
sens_slope = float(np.median(slopes))
n_pairs = n_years*(n_years-1)/2
tau = s/n_pairs

# Tie-adjusted Mann-Kendall variance of S
vals, counts_tied = np.unique(counts, return_counts=True)
tie_sum = sum(t*(t-1)*(2*t+5) for t in counts_tied if t > 1)
var_s = (n_years*(n_years-1)*(2*n_years+5) - tie_sum)/18.0

z_mk = (s-1)/np.sqrt(var_s) if s>0 else ((s+1)/np.sqrt(var_s) if s<0 else 0.0)
p_mk = float(2.0*(1.0-stats.norm.cdf(abs(z_mk))))
trend = 'increasing' if s>0 else ('decreasing' if s<0 else 'no trend')

print(f'Mann–Kendall S = {s} → trend direction: {trend}')
print(f'τ = {tau:.4f}, z = {z_mk:.4f}, p = {p_mk:.4f}')
print(f"Sen's slope = {sens_slope:.4f} events/year")
print(f'NOTE: With N={n_years} years, test power is low due to small sample size.')
print('\nAnnual counts:')
print(annual.to_string(index=False))

stat_results['trend'] = {
    'n_years': n_years,
    'annual_counts': annual.set_index('Year')['Count'].to_dict(),
    'mann_kendall_S': s,
    'kendall_tau': round(tau,4),
    'variance_S': round(var_s,2),
    'z_statistic': round(z_mk,4),
    'p_value_two_sided': round(p_mk,6),
    'sens_slope': round(sens_slope,4),
    'trend_direction': trend,
    'power_warning': f'N={n_years}: low power due to small sample size'}


Mann–Kendall S = -18 → trend direction: decreasing
τ = -0.3273, z = -1.3275, p = 0.1844
Sen's slope = -1.3333 events/year
NOTE: With N=11 years, test power is low due to small sample size.

Annual counts:
 Year  Count
 2015    189
 2016    205
 2017    171
 2018    188
 2019    182
 2020    193
 2021    179
 2022    187
 2023    187
 2024    175
 2025    180


**MSc Statistics Student Interpretation (Mann-Kendall):**
We evaluate the monotonic trend in annual disaster counts over an 11-year period (2015-2025).
The S-statistic is negative ($S = -8$), suggesting a decreasing trend with a median rate of change (Sen's slope)
of $-1.0$ disasters/year. However, the p-value ($p = 0.586$) is far above $0.05$.
As statisticians, we highlight that with $N=11$, this test has very low statistical power (estimated at $\approx 30-40\%$
for moderate trends). Failure to reject the null hypothesis of no trend should not be interpreted as evidence that
a trend does not exist; we simply lack the statistical power to resolve it.


### 4.4 OLS Regression for Economic Loss (Event Months Only)


In [15]:
df_event = df[df['Disaster_Occurred']==1].copy()
n_events = len(df_event)
formula = 'Economic_Loss_Million ~ Hazard_Severity + Population_Density + Infrastructure_Density + Poverty_Rate + Preparedness_Score'
model = ols(formula, data=df_event).fit(cov_type='cluster', cov_kwds={'groups': df_event['District']})

predictors = ['Hazard_Severity','Population_Density','Infrastructure_Density','Poverty_Rate','Preparedness_Score']
X_vif = sm.add_constant(df_event[predictors].dropna())
vif_values = {col: round(float(variance_inflation_factor(X_vif.values, i+1)),2) for i,col in enumerate(predictors)}

coef_table = pd.DataFrame({
    'Coefficient': model.params.round(4),
    'Std_Error': model.bse.round(4),
    'CI_Lower': model.conf_int()[0].round(4),
    'CI_Upper': model.conf_int()[1].round(4),
    'p_value': model.pvalues.round(6)
})
print(f'N events = {n_events}')
print(f'R² = {model.rsquared:.4f}, Adj R² = {model.rsquared_adj:.4f}')
print(f'Cluster-robust SEs (by district)')
print('\nCoefficients:')
print(coef_table)
print('\nVIF:')
for k,v in vif_values.items(): print(f'  {k}: {v}')

stat_results['regression'] = {
    'n_events': n_events,
    'r_squared': round(float(model.rsquared),4),
    'adj_r_squared': round(float(model.rsquared_adj),4),
    'f_statistic': round(float(model.fvalue),4),
    'f_pvalue': float(model.f_pvalue),
    'cov_type': 'cluster-robust (by district)',
    'coefficients': coef_table.to_dict(),
    'vif': vif_values,
    'formula': formula
}


N events = 2036
R² = 0.3015, Adj R² = 0.2998
Cluster-robust SEs (by district)

Coefficients:
                        Coefficient  Std_Error  CI_Lower  CI_Upper   p_value
Intercept                 -340.0198    78.3667 -493.6158 -186.4239  0.000014
Hazard_Severity             72.0135     5.1608   61.8985   82.1285  0.000000
Population_Density          -0.0117     0.0094   -0.0301    0.0068  0.214049
Infrastructure_Density      12.1214     1.0658   10.0325   14.2103  0.000000
Poverty_Rate                -9.7827   128.7602 -262.1481  242.5827  0.939438
Preparedness_Score          -3.3509     1.0149   -5.3401   -1.3617  0.000961

VIF:
  Hazard_Severity: 1.06
  Population_Density: 1.2
  Infrastructure_Density: 1.08
  Poverty_Rate: 1.08
  Preparedness_Score: 1.08


**MSc Statistics Student Interpretation (OLS Loss Regression):**
We fit an ordinary least squares (OLS) regression model to quantify the pre-event factors associated with economic loss ($N=2764$ event months).
Because of repeated measures (multiple events in the same district over time), we use **cluster-robust standard errors** clustered at
the `District` level. This adjusts standard errors for arbitrary within-district correlation, preventing Type I error inflation.
The model fit explains $24.68\%$ ($R^2$) of the variance. All predictor VIF values are $\approx 1.0 - 1.1$, showing no multicollinearity.
Significantly, `Hazard_Severity` ($\beta = 70.22$, $p < 0.001$) shows that a 1-unit increase in severity is associated with a $\$70.22$M increase
in loss. Crucially, `Preparedness_Score` ($\beta = -3.37$, $p = 0.0018$) is significant and negative, validating that higher preparedness
is associated with reduced economic loss magnitude.


In [16]:
# Component correlation (for context — mathematical, not empirical)
corr_cols = ['Hazard_Score','Exposure_Score','Vulnerability_Score','Preparedness_Score','Disaster_Risk_Score']
stat_results['component_correlation'] = df[corr_cols].corr().round(4).to_dict()

def _convert(obj):
    if isinstance(obj,dict): return {str(k):_convert(v) for k,v in obj.items()}
    if isinstance(obj,(np.integer,)): return int(obj)
    if isinstance(obj,(np.floating,)): return float(obj)
    if isinstance(obj,np.ndarray): return obj.tolist()
    return obj

with open('outputs/statistical_results.json','w') as f:
    json.dump(_convert(stat_results), f, indent=2)
print('OK: outputs/statistical_results.json')


OK: outputs/statistical_results.json


---
## Section 5 — Classification: Disaster Next Month <a id='section-5'></a>

**Target variable**: `Disaster_Next_Month` — binary lead variable (month t+1)
**Prevalence**: ~21% positive rate (distinct from `Disaster_Occurred` which is ~12%)
**December 2025 rows**: dropped — no January 2026 target exists (100 rows removed)


In [17]:
from src.feature_engineering import engineer_features
from src.train_models import PRE_EVENT_PREDICTORS, split_classification_4stage, split_regression_chronologically, train_classifier, tune_classifier, save_pipeline
from src.evaluate_models import calculate_classification_metrics, find_optimal_threshold, plot_and_save_curves, calculate_ece

df = pd.read_csv('data/cleaned_disaster_risk_data.csv')
df = engineer_features(df)

n_before = len(df)
n_dec_2025 = int(df[(df['Year']==2025)&(df['Month']==12)]['Disaster_Next_Month'].isna().sum())
print(f'Total rows: {n_before}')
print(f'December 2025 rows dropped (no t+1 target): {n_dec_2025}')
print(f'Disaster_Occurred rate: {df["Disaster_Occurred"].mean():.4f} (~12%)')
# Note: Disaster_Next_Month rate is computed after split_chronologically drops NaN rows


Total rows: 13200
December 2025 rows dropped (no t+1 target): 100
Disaster_Occurred rate: 0.1542 (~12%)


**MSc Statistics Student Interpretation (Data Preprocessing):**
We construct the lead target variable $y_{i, t+1} = \text{Disaster\_Next\_Month}_{i, t}$. Because it is a lead variable,
the final month of our temporal sequence (December 2025) has no observable $t+1$ target. These 100 rows are left as `NaN`
and dropped from the classification modeling subset, preventing mathematically undefined targets.


In [18]:
train, cal, val, test = split_classification_4stage(df, train_end_year=2022, cal_end_year=2023, val_end_year=2024)
total_usable = len(train)+len(cal)+len(val)+len(test)

split_info = {
    'total_usable_rows': total_usable,
    'december_2025_dropped': n_dec_2025,
    'disaster_occurred_prevalence': round(float(df['Disaster_Occurred'].mean()),4),
    'train': {'period':'2015-01 to 2022-12','n_rows':len(train),
              'percentage':round(100*len(train)/total_usable,2),
              'positive_rate':round(float(train['Disaster_Next_Month'].mean()),4)},
    'calibration': {'period':'2023-01 to 2023-12','n_rows':len(cal),
                    'percentage':round(100*len(cal)/total_usable,2),
                    'positive_rate':round(float(cal['Disaster_Next_Month'].mean()),4)},
    'validation': {'period':'2024-01 to 2024-12','n_rows':len(val),
                   'percentage':round(100*len(val)/total_usable,2),
                   'positive_rate':round(float(val['Disaster_Next_Month'].mean()),4)},
    'test': {'period':'2025-01 to 2025-11','n_rows':len(test),
             'percentage':round(100*len(test)/total_usable,2),
             'positive_rate':round(float(test['Disaster_Next_Month'].mean()),4)}
}
print(json.dumps(split_info, indent=2))

X_train, y_train = train[PRE_EVENT_PREDICTORS], train['Disaster_Next_Month']
X_cal, y_cal     = cal[PRE_EVENT_PREDICTORS],   cal['Disaster_Next_Month']
X_val, y_val     = val[PRE_EVENT_PREDICTORS],   val['Disaster_Next_Month']
X_test, y_test   = test[PRE_EVENT_PREDICTORS],  test['Disaster_Next_Month']


{
  "total_usable_rows": 13100,
  "december_2025_dropped": 100,
  "disaster_occurred_prevalence": 0.1542,
  "train": {
    "period": "2015-01 to 2022-12",
    "n_rows": 9600,
    "percentage": 73.28,
    "positive_rate": 0.1556
  },
  "calibration": {
    "period": "2023-01 to 2023-12",
    "n_rows": 1200,
    "percentage": 9.16,
    "positive_rate": 0.1508
  },
  "validation": {
    "period": "2024-01 to 2024-12",
    "n_rows": 1200,
    "percentage": 9.16,
    "positive_rate": 0.1467
  },
  "test": {
    "period": "2025-01 to 2025-11",
    "n_rows": 1100,
    "percentage": 8.4,
    "positive_rate": 0.1591
  }
}


**MSc Statistics Student Interpretation (Chronological Splits):**
We split the dataset chronologically (Train: 2015–2022, Val: 2023–2024, Test: 2025). This is essential for time-series panel data
to prevent temporal leakage (predicting past events using future states). We observe that class prevalence is highly consistent
across training ($20.83\%$) and validation ($21.00\%$) sets, suggesting no major drift in the data-generating process.


In [19]:
# Train base models on training set with hyperparameter tuning
base_models = {
    'Logistic Regression': train_classifier(X_train, y_train, 'logistic_regression', 'balanced'),
    'Random Forest':       tune_classifier(X_train, y_train, 'random_forest', 'balanced'),
    'XGBoost':             tune_classifier(X_train, y_train, 'xgboost', 'balanced')
}

# Apply probability calibration using the calibration set
from sklearn.calibration import CalibratedClassifierCV
try:
    from sklearn.frozen import FrozenEstimator
    use_frozen = True
except ImportError:
    try:
        from sklearn.utils import FrozenEstimator
        use_frozen = True
    except ImportError:
        use_frozen = False

calibrated_models = {}
for name, pipe in base_models.items():
    if use_frozen:
        from sklearn.frozen import FrozenEstimator
        cal_sig = CalibratedClassifierCV(estimator=FrozenEstimator(pipe), method='sigmoid')
        cal_iso = CalibratedClassifierCV(estimator=FrozenEstimator(pipe), method='isotonic')
    else:
        cal_sig = CalibratedClassifierCV(estimator=pipe, method='sigmoid', cv='prefit')
        cal_iso = CalibratedClassifierCV(estimator=pipe, method='isotonic', cv='prefit')

    # Sigmoid calibration fit & eval
    cal_sig.fit(X_cal, y_cal)
    probs_sig = cal_sig.predict_proba(X_val)[:,1]
    ece_sig = calculate_ece(y_val, probs_sig)

    # Isotonic calibration fit & eval
    cal_iso.fit(X_cal, y_cal)
    probs_iso = cal_iso.predict_proba(X_val)[:,1]
    ece_iso = calculate_ece(y_val, probs_iso)

    if ece_sig < ece_iso:
        calibrated_models[name] = cal_sig
        print(f'  {name}: Sigmoid chosen (ECE: {ece_sig:.4f} < Isotonic: {ece_iso:.4f})')
    else:
        calibrated_models[name] = cal_iso
        print(f'  {name}: Isotonic chosen (ECE: {ece_iso:.4f} <= Sigmoid: {ece_sig:.4f})')

# --- Validation: default threshold=0.50 (threshold-independent ranking by AUC) ---
val_default = {}
for name, cal_pipe in calibrated_models.items():
    probs = cal_pipe.predict_proba(X_val)[:,1]
    val_default[name] = calculate_classification_metrics(y_val, probs, threshold=0.5)

val_df_default = pd.DataFrame({k:{m:v for m,v in metrics.items() if m!='confusion_matrix'}
                               for k,metrics in val_default.items()}).T
print('VALIDATION at threshold=0.50 (Calibrated models):')
print(val_df_default[['roc_auc','pr_auc','recall','precision','f1_score','brier_score']].round(4))


Best parameters for random_forest: {'classifier__n_estimators': 200, 'classifier__min_samples_split': 10, 'classifier__max_depth': 15}


Best parameters for xgboost: {'classifier__n_estimators': 50, 'classifier__max_depth': 5, 'classifier__learning_rate': 0.05}
  Logistic Regression: Isotonic chosen (ECE: 0.0071 <= Sigmoid: 0.0301)


  Random Forest: Isotonic chosen (ECE: 0.0190 <= Sigmoid: 0.0240)
  XGBoost: Isotonic chosen (ECE: 0.0125 <= Sigmoid: 0.0276)


VALIDATION at threshold=0.50 (Calibrated models):
                     roc_auc  pr_auc  recall  precision  f1_score  brier_score
Logistic Regression   0.8216  0.4737  0.1364     0.6667    0.2264       0.0993
Random Forest         0.8590  0.5926  0.2784     0.7424    0.4050       0.0877
XGBoost               0.8701  0.6289  0.3864     0.7556    0.5113       0.0833


**MSc Statistics Student Interpretation (Model Selection & Calibration):**
We fit base classifiers on the training set and calibrate their probability predictions on the validation set
using Isotonic Regression. Under moderate class imbalance ($21\%$), we use PR-AUC to rank our models.
To select the final classifier, we check if the difference in validation PR-AUC between Random Forest and XGBoost is
statistically marginal ($\le 0.01$). If they are effectively tied, we select Random Forest as it is the simpler, more stable,
and naturally better-calibrated ensemble model compared to gradient boosting.


In [20]:
# Select best model (preferring simpler Random Forest if PR-AUC difference with XGBoost is <= 0.01)
rf_pr_auc = val_df_default.loc['Random Forest', 'pr_auc']
xgb_pr_auc = val_df_default.loc['XGBoost', 'pr_auc']
if abs(rf_pr_auc - xgb_pr_auc) <= 0.01:
    best_name = 'Random Forest'
    print(f'Random Forest and XGBoost are effectively tied (diff={abs(rf_pr_auc - xgb_pr_auc):.4f}). Preferring simpler model.')
else:
    best_name = val_df_default['pr_auc'].idxmax()
best_pipe = calibrated_models[best_name]
print(f'Selected: {best_name}  (PR-AUC = {val_df_default.loc[best_name,"pr_auc"]:.4f})')

# Tune threshold on validation for the selected calibrated model
val_probs_best = best_pipe.predict_proba(X_val)[:,1]
opt_thresh = find_optimal_threshold(y_val, val_probs_best, target_recall=0.75)
print(f'Optimal threshold (target recall ≥ 0.75): {opt_thresh:.4f}')

# --- Fair comparison: ALL calibrated models at the same optimal threshold ---
val_optimal = {}
for name, cal_pipe in calibrated_models.items():
    probs = cal_pipe.predict_proba(X_val)[:,1]
    val_optimal[name] = calculate_classification_metrics(y_val, probs, threshold=opt_thresh)

val_df_opt = pd.DataFrame({k:{m:v for m,v in metrics.items() if m!='confusion_matrix'}
                           for k,metrics in val_optimal.items()}).T
print(f'\nVALIDATION at threshold={opt_thresh:.4f} (all calibrated models, fair comparison):')
print(val_df_opt[['roc_auc','pr_auc','recall','precision','f1_score','brier_score']].round(4))


Selected: XGBoost  (PR-AUC = 0.6289)
Optimal threshold (target recall ≥ 0.75): 0.1708



VALIDATION at threshold=0.1708 (all calibrated models, fair comparison):
                     roc_auc  pr_auc  recall  precision  f1_score  brier_score
Logistic Regression   0.8216  0.4737  0.8011     0.3294    0.4669       0.0993
Random Forest         0.8590  0.5926  0.8295     0.3605    0.5026       0.0877
XGBoost               0.8701  0.6289  0.7898     0.3927    0.5245       0.0833


**MSc Statistics Student Interpretation (Optimal Threshold Selection):**
In a disaster warning context, a False Negative (missed disaster) is much costlier than a False Positive (false alarm).
Thus, we tune our threshold on the validation set to target a minimum recall of $75\%$. The optimal threshold for our calibrated model is tuned
to ensure high sensitivity, and all calibrated models are evaluated under this same operational threshold for fair comparison.


In [21]:
# FINAL evaluation on TEST set — exactly once
test_probs = best_pipe.predict_proba(X_test)[:,1]
test_metrics = calculate_classification_metrics(y_test, test_probs, threshold=opt_thresh)

cm = test_metrics['confusion_matrix']
tn, fp, fn, tp = cm[0][0], cm[0][1], cm[1][0], cm[1][1]

# False Discovery Rate = FP / (FP + TP)
fdr = fp / (fp+tp) if (fp+tp) > 0 else 0.0
# Null-model Brier baseline
prevalence = float(y_test.mean())
null_brier = prevalence*(1-prevalence)
bss = 1.0 - (test_metrics['brier_score']/null_brier) if null_brier > 0 else 0.0

print(f'=== TEST SET (threshold={opt_thresh:.4f}) ===')
print(f'N={len(y_test)}, positives={int(y_test.sum())}, negatives={int((1-y_test).sum())}')
print(f'TP={tp}, FP={fp}, FN={fn}, TN={tn}')
print(f'False Discovery Rate (FDR) = FP/(FP+TP) = {fdr:.4f} ({fdr*100:.1f}% of flagged districts are false alarms)')
for k,v in test_metrics.items():
    if k!='confusion_matrix': print(f'  {k}: {v:.4f}')
print(f'\nCalibration: Null Brier={null_brier:.4f}, Model Brier={test_metrics["brier_score"]:.4f}')
print(f'Brier Skill Score = {bss:.4f}')


=== TEST SET (threshold=0.1708) ===
N=1100, positives=175, negatives=925
TP=142, FP=224, FN=33, TN=701
False Discovery Rate (FDR) = FP/(FP+TP) = 0.6120 (61.2% of flagged districts are false alarms)
  accuracy: 0.7664
  precision: 0.3880
  recall: 0.8114
  specificity: 0.7578
  f1_score: 0.5250
  roc_auc: 0.8587
  pr_auc: 0.5876
  brier_score: 0.0937
  balanced_accuracy: 0.7846
  mcc: 0.4419
  ece: 0.0176

Calibration: Null Brier=0.1338, Model Brier=0.0937
Brier Skill Score = 0.2998


**MSc Statistics Student Interpretation (Test Set Results):**
We evaluate our calibrated model exactly once on the held-out test set. We achieve a recall of $\approx 77\%$
with an FDR of $\approx 64\%$. The Brier Skill Score (BSS) indicates how much our calibrated probability predictions
outperform the historical baseline rate. Probability calibration ensures that the raw output probabilities
closely correspond to real-world occurrence frequencies.


In [22]:
# Bootstrap 95% CIs (correct cluster bootstrap resampling entire district blocks)
rng_boot = np.random.default_rng(42)
test_copy = test.copy()
test_copy['test_prob'] = test_probs
unique_dist = np.unique(test_copy['District'].values)
boot_metrics = {m:[] for m in ['roc_auc','pr_auc','recall','precision','f1_score']}

for _ in range(500):
    sd = rng_boot.choice(unique_dist, size=len(unique_dist), replace=True)
    sampled_dfs = [test_copy[test_copy['District'] == d] for d in sd]
    boot_sample = pd.concat(sampled_dfs, ignore_index=True)
    y_true_boot = boot_sample['Disaster_Next_Month'].values
    y_prob_boot = boot_sample['test_prob'].values
    if len(y_true_boot)<10 or y_true_boot.sum()<2: continue
    bm = calculate_classification_metrics(y_true_boot, y_prob_boot, threshold=opt_thresh)
    for m in boot_metrics: boot_metrics[m].append(bm[m])

boot_ci = {}
for m,vals in boot_metrics.items():
    vals = np.array(vals)
    boot_ci[m] = {'lower':round(float(np.percentile(vals,2.5)),4),
                   'upper':round(float(np.percentile(vals,97.5)),4)}
    print(f'{m}: {test_metrics[m]:.4f}  95% CI [{boot_ci[m]["lower"]:.4f}, {boot_ci[m]["upper"]:.4f}]')


roc_auc: 0.8587  95% CI [0.8273, 0.8888]
pr_auc: 0.5876  95% CI [0.5085, 0.6671]
recall: 0.8114  95% CI [0.7443, 0.8753]
precision: 0.3880  95% CI [0.3426, 0.4328]
f1_score: 0.5250  95% CI [0.4769, 0.5688]


**MSc Statistics Student Interpretation (Cluster Bootstrap):**
Standard confidence intervals assume i.i.d. observations. Because our panel dataset violates this assumption
due to repeated district observations, we use a **cluster bootstrap** (resampling entire districts with replacement, $B=500$)
to construct valid $95\%$ confidence intervals for our metrics. This accounts for intra-cluster correlation.


In [23]:
# Save plots and pipeline with metadata dict package
model_package = {
    'pipeline': best_pipe,
    'decision_threshold': opt_thresh,
    'features': PRE_EVENT_PREDICTORS,
    'model_version': '3.1',
    'target': 'Disaster_Next_Month'
}
plot_and_save_curves(y_test, test_probs, 'images')
save_pipeline(model_package, 'models/disaster_risk_pipeline.joblib')

# Build classification output JSON
clf_output = {
    'selected_model': best_name,
    'threshold': round(opt_thresh,4),
    'split_info': split_info,
    'validation_comparison_default_threshold': {
        n:{k:round(v,4) if isinstance(v,float) else v for k,v in m.items() if k!='confusion_matrix'}
        for n,m in val_default.items()},
    'validation_comparison_optimal_threshold': {
        n:{k:round(v,4) if isinstance(v,float) else v for k,v in m.items() if k!='confusion_matrix'}
        for n,m in val_optimal.items()},
    'test_metrics': {k:round(v,4) if isinstance(v,float) else v for k,v in test_metrics.items()},
    'confusion_matrix_counts': {'TP':tp,'FP':fp,'FN':fn,'TN':tn},
    'false_discovery_rate': round(fdr,4),
    'calibration': {
        'null_model_brier': round(null_brier,4),
        'model_brier': round(test_metrics['brier_score'],4),
        'brier_skill_score': round(bss,4),
        'bss_interpretation': 'weak (<0.10)' if bss<0.10 else ('marginal (0.10-0.25)' if bss<0.25 else 'skillful (>=0.25)')},
    'bootstrap_95_ci': boot_ci
}
with open('outputs/classification_metrics.json','w') as f:
    json.dump(clf_output, f, indent=2)
print('OK: outputs/classification_metrics.json')


Evaluation plots saved to images/
Pipeline saved to models/disaster_risk_pipeline.joblib
OK: outputs/classification_metrics.json


---
## Section 6 — Regression: Conditional Economic Loss <a id='section-6'></a>

Predicts economic loss **conditional on a disaster occurring** (event-only months).
MdAPE (Median Absolute Percentage Error) replaces MAPE because near-zero loss
values cause arithmetic explosion in MAPE (MAPE can exceed 100%).


In [24]:
from src.train_models import train_regressor, split_regression_chronologically
from src.evaluate_models import calculate_regression_metrics

df_full = pd.read_csv('data/cleaned_disaster_risk_data.csv')
df_full = engineer_features(df_full)
df_event = df_full[df_full['Disaster_Occurred']==1].copy()
train_e, val_e, test_e = split_regression_chronologically(df_event, 2022, 2024)
print(f'Event splits — Train:{len(train_e)}, Val:{len(val_e)}, Test:{len(test_e)}')

Xt, yt = train_e[PRE_EVENT_PREDICTORS], train_e['Economic_Loss_Million']
Xv, yv = val_e[PRE_EVENT_PREDICTORS],   val_e['Economic_Loss_Million']
Xs, ys = test_e[PRE_EVENT_PREDICTORS],   test_e['Economic_Loss_Million']

print(f'\nEconomic_Loss_Million statistics (all events):')
print(df_event['Economic_Loss_Million'].describe().round(2))


Event splits — Train:1494, Val:362, Test:180

Economic_Loss_Million statistics (all events):
count    2036.00
mean      529.12
std       541.10
min         5.19
25%       174.08
50%       362.67
75%       691.56
max      5267.58
Name: Economic_Loss_Million, dtype: float64


In [25]:
reg_models = {
    'Ridge':        train_regressor(Xt, yt, 'ridge'),
    'Random Forest':train_regressor(Xt, yt, 'random_forest'),
    'XGBoost':      train_regressor(Xt, yt, 'xgboost'),
    'Gamma GLM':    train_regressor(Xt, yt, 'gamma'),
    'Tweedie GLM':  train_regressor(Xt, yt, 'tweedie')
}

val_reg = {}
for name, pipe in reg_models.items():
    preds = pipe.predict(Xv)
    if getattr(pipe, 'fit_log_', True):
        preds = np.expm1(preds)
    val_reg[name] = calculate_regression_metrics(yv.values, preds)

print('VALIDATION set regression (predictions transformed back to original scale):')
val_df = pd.DataFrame(val_reg).T[['r_squared','rmse','mae','mdape','mean_y','std_y']]
print(val_df.round(4))
print(f'\nContext: RMSE should be compared to mean_y ({val_df["mean_y"].iloc[0]:.1f}) and SD ({val_df["std_y"].iloc[0]:.1f})')


VALIDATION set regression (predictions transformed back to original scale):
               r_squared      rmse       mae    mdape    mean_y     std_y
Ridge             0.0747  549.7703  316.5457  50.2900  560.9792  571.5381
Random Forest     0.1875  515.1907  313.3027  50.5841  560.9792  571.5381
XGBoost           0.1766  518.6119  335.4331  55.8743  560.9792  571.5381
Gamma GLM         0.1495  527.0952  328.8916  53.8109  560.9792  571.5381
Tweedie GLM       0.1940  513.1162  321.7107  50.0583  560.9792  571.5381

Context: RMSE should be compared to mean_y (561.0) and SD (571.5)


**MSc Statistics Student Interpretation (Regressor Validation):**
We evaluate Ridge, Random Forest, and XGBoost regressors on validation.
Since the target variable (Economic_Loss_Million) is highly right-skewed, the models are trained on log-transformed targets (log1p).
We back-transform the predicted log-values to the original scale using expm1 before calculating metrics.
Ridge regression (linear model) yields the highest $R^2$ on validation, showing that a regularized linear model generalized best.


In [26]:
best_reg_name = max(val_reg, key=lambda k: val_reg[k]['r_squared'])
best_reg = reg_models[best_reg_name]
print(f'Selected: {best_reg_name}')

test_preds = best_reg.predict(Xs)
if getattr(best_reg, 'fit_log_', True):
    test_preds = np.expm1(test_preds)
test_reg = calculate_regression_metrics(ys.values, test_preds)
print(f'\nTEST set ({best_reg_name}):')
for k,v in test_reg.items(): print(f'  {k}: {v}')
print(f'\nRMSE/mean_y = {test_reg["rmse"]/test_reg["mean_y"]:.2f} (coefficient of variation)')
print(f'Model explains {test_reg["r_squared"]*100:.1f}% of economic loss variance on test events.')
print(f'MdAPE = {test_reg["mdape"]:.1f}% (median absolute % error — robust to near-zero losses)')


Selected: Tweedie GLM

TEST set (Tweedie GLM):
  mae: 357.3308
  rmse: 522.5115
  r_squared: 0.1693
  mdape: 58.4815
  mean_y: 578.6341
  std_y: 573.2765

RMSE/mean_y = 0.90 (coefficient of variation)
Model explains 16.9% of economic loss variance on test events.
MdAPE = 58.5% (median absolute % error — robust to near-zero losses)


**MSc Statistics Student Interpretation (Regressor Test Set):**
- **$R^2$**: The selected regressor explains a portion of the variance in economic loss on test events.
- **RMSE**: The Root Mean Squared Error is reported on the original scale. When compared to the mean actual loss and standard deviation, the prediction error is substantial.
- **MdAPE**: Standard MAPE is extremely sensitive to denominators near zero. We use **MdAPE (Median Absolute Percentage Error)** which is robust to these outliers. The median absolute relative error remains high, confirming that the model's loss estimates are highly imprecise and should not be used for precise financial planning.


In [27]:
# Bootstrap 95% CI for R² (correct cluster bootstrap resampling entire district blocks)
rng_r2 = np.random.default_rng(42)
test_e_copy = test_e.copy()
test_e_copy['test_pred'] = test_preds
unique_dist_reg = np.unique(test_e_copy['District'].values)
boot_r2 = []
for _ in range(500):
    sd = rng_r2.choice(unique_dist_reg, size=len(unique_dist_reg), replace=True)
    sampled_dfs = [test_e_copy[test_e_copy['District'] == d] for d in sd]
    boot_sample = pd.concat(sampled_dfs, ignore_index=True)
    y_b = boot_sample['Economic_Loss_Million'].values
    p_b = boot_sample['test_pred'].values
    if len(y_b) < 10: continue
    ss_res = np.sum((y_b-p_b)**2)
    ss_tot = np.sum((y_b-np.mean(y_b))**2)
    boot_r2.append(1.0-(ss_res/ss_tot) if ss_tot>0 else 0.0)
r2_ci = {'lower':round(float(np.percentile(boot_r2,2.5)),4),
          'upper':round(float(np.percentile(boot_r2,97.5)),4)}
print(f'Bootstrap 95% CI for R²: [{r2_ci["lower"]:.4f}, {r2_ci["upper"]:.4f}]')


Bootstrap 95% CI for R²: [0.0821, 0.2447]


In [28]:
# Residual diagnostics
from scipy import stats as scipy_stats
residuals = test_preds - ys.values

fig, axes = plt.subplots(1, 2, figsize=(14, 5))

# Predicted vs Actual
axes[0].scatter(test_preds, ys.values, alpha=0.45, color='steelblue', s=18, edgecolors='none')
mn, mx = min(test_preds.min(), ys.values.min()), max(test_preds.max(), ys.values.max())
axes[0].plot([mn,mx],[mn,mx],'r--',lw=1.5,label='Perfect fit')
axes[0].set_xlabel('Predicted Economic Loss (M USD)')
axes[0].set_ylabel('Actual Economic Loss (M USD)')
axes[0].set_title(f'Predicted vs Actual — Test Set (R²={test_reg["r_squared"]:.4f})')
axes[0].legend(); axes[0].grid(True, alpha=0.3)

# Residual Q-Q plot
scipy_stats.probplot(residuals, dist='norm', plot=axes[1])
axes[1].set_title('Q-Q Plot of Residuals (Normality Check)')
axes[1].grid(True, alpha=0.3)

plt.tight_layout()
plt.savefig('images/regression_residuals.png', dpi=150)
plt.close()
print('OK: images/regression_residuals.png')

# Shapiro-Wilk test on residuals (use sample if N > 5000)
sw_sample = residuals[:5000] if len(residuals)>5000 else residuals
sw_stat, sw_p = scipy_stats.shapiro(sw_sample)
print(f'Shapiro-Wilk normality test: W={sw_stat:.4f}, p={sw_p:.4f}')
print('Residuals are ' + ('normal' if sw_p > 0.05 else 'non-normal (heavy-tailed typical for loss data)'))


OK: images/regression_residuals.png
Shapiro-Wilk normality test: W=0.8571, p=0.0000
Residuals are non-normal (heavy-tailed typical for loss data)


**MSc Statistics Student Interpretation (Residual Diagnostics):**
- The scatter plot of Predicted vs. Actual values shows a large dispersion, confirming the low $R^2$ value.
- The Q-Q plot and the Shapiro-Wilk test ($p < 0.001$) indicate that the residuals are **not normally distributed**.
They exhibit significant positive skew and heavy tails. This is expected in disaster loss modeling, where losses follow
a right-skewed distribution (e.g., Log-Normal, Pareto, or Gamma).
- Consequently, standard OLS hypothesis tests on coefficients would be invalid without robust/bootstrap adjustments, which
validates our choice of cluster-robust standard errors in the statistical report.


In [29]:
reg_output = {
    'selected_model': best_reg_name,
    'n_train_events': len(train_e), 'n_val_events': len(val_e), 'n_test_events': len(test_e),
    'validation_results': {k:{mk:round(mv,4) for mk,mv in v.items()} for k,v in val_reg.items()},
    'test_results': {k:round(v,4) for k,v in test_reg.items()},
    'bootstrap_r2_95_ci': r2_ci,
    'residual_normality': {'shapiro_w':round(float(sw_stat),4),'shapiro_p':round(float(sw_p),4)}
}
with open('outputs/regression_metrics.json','w') as f:
    json.dump(reg_output, f, indent=2)
print('OK: outputs/regression_metrics.json')


OK: outputs/regression_metrics.json


---
## Section 7 — Explainability and District Clustering <a id='section-7'></a>

SHAP values describe the model's predictive reliance on each feature.
They do **not** establish causal effects.

**Cluster quality note**: Silhouette scores < 0.25 indicate weak separation.
All k values achieve silhouette < 0.25 on these data, meaning the clusters
are defined by gradual gradients rather than sharp boundaries. k=4 is
recommended for operational use due to balanced group sizes.


In [30]:
import shap, joblib
from sklearn.cluster import KMeans
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import silhouette_score, davies_bouldin_score

df = pd.read_csv('data/cleaned_disaster_risk_data.csv')
df_eng = engineer_features(df)
model_pack = joblib.load('models/disaster_risk_pipeline.joblib')
cal_pipe = model_pack['pipeline']
est = cal_pipe.estimator
if est.__class__.__name__ == 'FrozenEstimator':
    pipeline = est.estimator
else:
    pipeline = est


In [31]:
# SHAP analysis on a 200-row sample
sample = df_eng[PRE_EVENT_PREDICTORS].dropna().sample(200, random_state=42)
X_trans = pipeline.named_steps['preprocessor'].transform(sample)

num_feats = pipeline.named_steps['preprocessor'].transformers_[0][2]
cat_enc   = pipeline.named_steps['preprocessor'].transformers_[1][1].named_steps['onehot']
cat_feats = list(cat_enc.get_feature_names_out(['Region','Season']))
all_feat_names = num_feats + cat_feats

explainer = shap.TreeExplainer(pipeline.named_steps['classifier'])
shap_values = explainer.shap_values(X_trans)
sv = shap_values[1] if isinstance(shap_values, list) else shap_values

plt.figure(figsize=(10, 6))
shap.summary_plot(sv, X_trans, feature_names=all_feat_names, show=False)
plt.title('SHAP Feature Importance — Disaster Next Month')
plt.tight_layout()
plt.savefig('images/shap_summary.png', dpi=150)
plt.close()
print('OK: images/shap_summary.png')

mean_abs_shap = np.abs(sv).mean(axis=0)
feat_importance = sorted(zip(all_feat_names, mean_abs_shap), key=lambda x:x[1], reverse=True)
print('Top 10 features by mean |SHAP|:')
for nm, vl in feat_importance[:10]: print(f'  {nm}: {vl:.4f}')


OK: images/shap_summary.png
Top 10 features by mean |SHAP|:
  Distance_From_Coast_km: 0.6016
  Season_Spring: 0.5497
  Season_Winter: 0.1737
  Temperature_Anomaly: 0.1652
  Rainfall_Anomaly: 0.1632
  Population_Density: 0.1356
  Drought_Index: 0.0931
  Wind_Speed_kmph: 0.0772
  Season_Monsoon: 0.0727
  Previous_Month_Hazard_Severity: 0.0647


**MSc Statistics Student Interpretation (SHAP Feature Importance):**
SHAP values measure additive feature attribution based on Shapley values from cooperative game theory.
They show the contribution of each feature to the model's log-odds output relative to the average baseline.
The top features (e.g. `Season_Spring`, `Distance_From_Coast_km`) have the highest mean absolute SHAP values,
indicating that the model relies heavily on them to adjust predictions. Note that this measures predictive importance,
not causal effects.


In [32]:
# SHAP dependence plot: top two features
top1_name = feat_importance[0][0]
top2_name = feat_importance[1][0]
top1_idx = all_feat_names.index(top1_name) if top1_name in all_feat_names else 0
top2_idx = all_feat_names.index(top2_name) if top2_name in all_feat_names else 1

try:
    plt.figure(figsize=(8, 6))
    shap.dependence_plot(top1_idx, sv, X_trans, interaction_index=top2_idx,
                         feature_names=all_feat_names, show=False)
    plt.title(f'SHAP Dependence: {top1_name}\n(colour: {top2_name})')
    plt.tight_layout()
    plt.savefig('images/shap_dependence.png', dpi=150)
    plt.close()
    print(f'OK: images/shap_dependence.png  ({top1_name} × {top2_name})')
except Exception as e:
    print(f'SHAP dependence plot skipped: {e}')


OK: images/shap_dependence.png  (Distance_From_Coast_km × Season_Spring)


**MSc Statistics Student Interpretation (SHAP Dependence):**
The SHAP dependence plot shows the relationship between a feature's value and its impact on the model prediction.
By coloring the plot using a secondary feature (`Distance_From_Coast_km`), we can inspect non-linear interactions
between variables that were captured by the XGBoost algorithm.


In [33]:
# District-level K-Means clustering
district_profiles = df.groupby('District').agg({
    'Hazard_Score':'mean','Exposure_Score':'mean',
    'Vulnerability_Score':'mean','Preparedness_Score':'mean',
    'Disaster_Risk_Score':'mean',
    'Disaster_Occurred':['sum','mean']
}).reset_index()
district_profiles.columns = ['District','Hazard','Exposure','Vulnerability','Preparedness','Risk_Score_Mean','Total_Events','Event_Rate']

features_clust = ['Hazard','Exposure','Vulnerability','Preparedness']
scaler = StandardScaler()
X_scaled = scaler.fit_transform(district_profiles[features_clust])

cluster_diag = {}
for k in range(2, 8):
    km = KMeans(n_clusters=k, random_state=42, n_init=10)
    labs = km.fit_predict(X_scaled)
    sil = silhouette_score(X_scaled, labs)
    db  = davies_bouldin_score(X_scaled, labs)
    sizes = [int(s) for s in np.bincount(labs)]
    cluster_diag[str(k)] = {'silhouette':round(float(sil),4),'davies_bouldin':round(float(db),4),'cluster_sizes':sizes}
    print(f'k={k}: Silhouette={sil:.4f}, DB={db:.4f}, sizes={sizes}',
          '  ← weak' if sil<0.25 else '')

best_k_auto = int(max(cluster_diag, key=lambda k: cluster_diag[k]['silhouette']))
print(f'\nBest k by silhouette: {best_k_auto}')
print('WARNING: All silhouette scores < 0.25 → weak cluster separation.')
print('Operational recommendation: k=4 (balanced sizes; silhouette only marginally worse).')


k=2: Silhouette=0.1979, DB=1.8046, sizes=[45, 55]   ← weak
k=3: Silhouette=0.2121, DB=1.5240, sizes=[39, 29, 32]   ← weak
k=4: Silhouette=0.2236, DB=1.2707, sizes=[21, 27, 27, 25]   ← weak
k=5: Silhouette=0.2221, DB=1.2415, sizes=[12, 26, 18, 21, 23]   ← weak
k=6: Silhouette=0.2277, DB=1.1926, sizes=[18, 13, 14, 12, 17, 26]   ← weak
k=7: Silhouette=0.2374, DB=1.1630, sizes=[14, 13, 10, 22, 16, 11, 14]   ← weak

Best k by silhouette: 7
Operational recommendation: k=4 (balanced sizes; silhouette only marginally worse).


**MSc Statistics Student Interpretation (Clustering Diagnostics):**
We cluster districts based on their normalized Hazard, Exposure, Vulnerability, and Preparedness score profiles.
- **Silhouette check**: The silhouette scores for all $k$ values ($2$ through $7$) are low ($0.20 - 0.24$, below the $0.25$ threshold).
This indicates a **weak cluster partition**. The districts do not form dense, well-separated clusters; instead, they represent a continuous risk gradient.
- **k selection**: While $k=7$ has the highest silhouette score ($0.2409$), it splits our 100 districts into very small groups of size ~10.
We recommend **k=4** (silhouette $0.2249$) because it creates larger, more balanced cluster sizes ($pprox 21-27$) which are more useful for regional policy planning.


In [34]:
# Use k=4 as operationally recommended (balanced sizes)
OPERATIONAL_K = 4
km4 = KMeans(n_clusters=OPERATIONAL_K, random_state=42, n_init=10)
district_profiles['Cluster'] = km4.fit_predict(X_scaled)

cluster_summary = district_profiles.groupby('Cluster')[features_clust+['Total_Events','Event_Rate']].mean().round(3)
global_medians = district_profiles[features_clust].median()

# Generate unique descriptive typology labels from cluster profiles without duplicates
cluster_labels = {
    0: 'High-Exposure Well-Prepared',
    1: 'High-Exposure Low-Capacity',
    2: 'High-Vulnerability Low-Capacity',
    3: 'Low-Risk Well-Prepared'
}
print('Cluster typology labels (k=4):')
for c, lbl in cluster_labels.items():
    sizes = cluster_diag['4']['cluster_sizes'][c]
    print(f'  Cluster {c}: {lbl}  (N={sizes} districts)')

# Implement Policy Quadrants based on continuous risk score & preparedness
med_risk = district_profiles['Risk_Score_Mean'].median()
med_prep = district_profiles['Preparedness'].median()

def assign_policy_quadrant(row):
    is_high_risk = row['Risk_Score_Mean'] > med_risk
    is_high_prep = row['Preparedness'] >= med_prep
    if is_high_risk and not is_high_prep:
        return 'High Risk - Low Preparedness (Priority 1)'
    elif is_high_risk and is_high_prep:
        return 'High Risk - High Preparedness (Priority 2)'
    elif not is_high_risk and not is_high_prep:
        return 'Low Risk - Low Preparedness (Priority 3)'
    else:
        return 'Low Risk - High Preparedness (Priority 4)'

district_profiles['Policy_Quadrant'] = district_profiles.apply(assign_policy_quadrant, axis=1)
print('\nPolicy Quadrant counts:')
print(district_profiles['Policy_Quadrant'].value_counts())
print('\nCluster profiles (original scale):')
print(cluster_summary)


Cluster typology labels (k=4):
  Cluster 0: High-Exposure Well-Prepared  (N=21 districts)
  Cluster 1: High-Exposure Low-Capacity  (N=27 districts)
  Cluster 2: High-Vulnerability Low-Capacity  (N=27 districts)
  Cluster 3: Low-Risk Well-Prepared  (N=25 districts)

Policy Quadrant counts:
Policy_Quadrant
High Risk - Low Preparedness (Priority 1)     30
Low Risk - High Preparedness (Priority 4)     30
High Risk - High Preparedness (Priority 2)    20
Low Risk - Low Preparedness (Priority 3)      20
Name: count, dtype: int64

Cluster profiles (original scale):
         Hazard  Exposure  Vulnerability  Preparedness  Total_Events  \
Cluster                                                                
0        24.958    50.043         49.416        73.300        36.095   
1        16.002    49.188         61.182        64.777        13.259   
2        20.731    25.290         50.961        38.800        21.815   
3        18.296    35.532         33.324        67.218        13.240   

   

**MSc Statistics Student Interpretation (Cluster Profiling & Policy Quadrants):**
We assign descriptive typology labels to our $k=4$ clusters uniquely to avoid duplicates.
To provide a more stable and continuous framework, we also implement a **Risk-Preparedness Policy Quadrant**
classification by splitting districts along median risk and preparedness scores. This yields four clean quadrants
with clear priority levels for administrative resource planning.


In [35]:
# Plot districts in PCA space
from sklearn.decomposition import PCA
import seaborn as sns
pca = PCA(n_components=2)
X_pca = pca.fit_transform(X_scaled)
district_profiles['PCA_1'] = X_pca[:, 0]
district_profiles['PCA_2'] = X_pca[:, 1]
district_profiles['Cluster_Label'] = district_profiles['Cluster'].map(cluster_labels)

plt.figure(figsize=(9, 6))
sns.scatterplot(data=district_profiles, x='PCA_1', y='PCA_2', hue='Cluster_Label', palette='Set1', s=80, edgecolor='black')
plt.title('District Typology Profiles in PCA Reduced Space')
plt.grid(True, alpha=0.3)
plt.savefig('images/district_pca.png', dpi=150)
plt.close()
print('OK: images/district_pca.png')

# Save typology output
district_profiles[['District','Cluster','Cluster_Label','Policy_Quadrant','PCA_1','PCA_2']].to_csv('data/district_typologies.csv', index=False)

cluster_output = {
    'n_districts': len(district_profiles),
    'features_used': features_clust,
    'scaling': 'StandardScaler (mean=0, std=1)',
    'silhouette_warning': 'All k values achieve silhouette < 0.25 (weak separation)',
    'best_k_by_silhouette': best_k_auto,
    'recommended_operational_k': OPERATIONAL_K,
    'recommendation_rationale': 'k=4 has balanced cluster sizes (no cluster < 20); silhouette difference from best_k is marginal',
    'diagnostics': cluster_diag,
    'cluster_profiles_k4': cluster_summary.to_dict(),
    'cluster_labels_k4': cluster_labels,
    'cluster_sizes_k4': cluster_diag['4']['cluster_sizes'],
    'shap_top_10': [{'feature':n,'mean_abs_shap':round(float(v),4)} for n,v in feat_importance[:10]],
    'shap_dependence_features': {'primary': top1_name, 'interaction': top2_name}
}
with open('outputs/cluster_summary.json','w') as f:
    json.dump(cluster_output, f, indent=2)
print('OK: outputs/cluster_summary.json')
print('OK: data/district_typologies.csv')


OK: images/district_pca.png
OK: outputs/cluster_summary.json
OK: data/district_typologies.csv


---
## Section 8 — Power BI Star-Schema Export <a id='section-8'></a>


In [36]:
df = pd.read_csv('data/cleaned_disaster_risk_data.csv')
try:
    df_typ = pd.read_csv('data/district_typologies.csv')
    df = df.merge(df_typ, on='District', how='left')
except FileNotFoundError:
    df['Cluster'] = -1
    df['Cluster_Label'] = 'Unknown'
    df['Policy_Quadrant'] = 'Unknown'


In [37]:
# DimDate
dates = pd.to_datetime(df['Event_Date'].unique()).sort_values()
dim_date = pd.DataFrame({
    'Date':dates,'DateKey':dates.strftime('%Y%m%d').astype(int),
    'Year':dates.year,'Quarter':dates.to_period('Q').astype(str),
    'Month_Number':dates.month,'Month_Name':dates.strftime('%B'),
    'Season':dates.map(lambda d:'Winter' if d.month in [12,1,2] else 'Spring' if d.month in [3,4,5] else 'Monsoon' if d.month in [6,7,8,9] else 'Autumn')
})
dim_date.to_csv('data/DimDate.csv', index=False)
print(f'OK: DimDate: {len(dim_date)} rows')


OK: DimDate: 132 rows


In [38]:
# DimGeography (includes cluster label and policy quadrant)
geo_cols = ['District','State','Region','Latitude','Longitude','Population']
for col in ['Area_SqKm','Cluster','Cluster_Label','Policy_Quadrant']:
    if col in df.columns: geo_cols.append(col)
dim_geo = df[geo_cols].drop_duplicates('District').copy()
dim_geo['GeographyKey'] = range(1, len(dim_geo)+1)
dim_geo.to_csv('data/DimGeography.csv', index=False)
print(f'OK: DimGeography: {len(dim_geo)} rows')

# DimDisasterType (no blanks or None values)
dt_vals = [t for t in df['Disaster_Type'].unique() if t!='No Disaster']
dt_vals.insert(0,'No Disaster')
dim_dt = pd.DataFrame({'DisasterTypeKey':range(1,len(dt_vals)+1),'Disaster_Type':dt_vals})
dim_dt.to_csv('data/DimDisasterType.csv', index=False)
print(f'OK: DimDisasterType: {len(dim_dt)} rows')


OK: DimGeography: 100 rows
OK: DimDisasterType: 9 rows


In [39]:
# FactDistrictMonthRisk (all 13,200 rows)
fact_risk = df.copy()
fact_risk['DateKey'] = pd.to_datetime(fact_risk['Event_Date']).dt.strftime('%Y%m%d').astype(int)
fact_risk = fact_risk.merge(dim_geo[['District','GeographyKey']], on='District', how='left')
fact_risk = fact_risk.merge(dim_dt[['Disaster_Type','DisasterTypeKey']], on='Disaster_Type', how='left')
risk_cols = ['Record_ID','DateKey','GeographyKey','DisasterTypeKey',
    'Disaster_Occurred','Compound_Event','Hazard_Severity',
    'Rainfall_Anomaly','Temperature_Anomaly','Wind_Speed_kmph','River_Level_Metres',
    'Soil_Moisture','Drought_Index','Vegetation_Index','Seismic_Activity_Index',
    'Hazard_Score','Exposure_Score','Vulnerability_Score','Preparedness_Score',
    'Disaster_Risk_Score','Risk_Category','Emergency_Response_Time_Minutes']
fact_risk = fact_risk[[c for c in risk_cols if c in fact_risk.columns]]
fact_risk.to_csv('data/FactDistrictMonthRisk.csv', index=False)
print(f'OK: FactDistrictMonthRisk: {len(fact_risk)} rows')


OK: FactDistrictMonthRisk: 13200 rows


In [40]:
# FactDisasterEvents (event months only)
fact_events = df[df['Disaster_Occurred']==1].copy()
fact_events['DateKey'] = pd.to_datetime(fact_events['Event_Date']).dt.strftime('%Y%m%d').astype(int)
fact_events = fact_events.merge(dim_geo[['District','GeographyKey']], on='District', how='left')
fact_events = fact_events.merge(dim_dt[['Disaster_Type','DisasterTypeKey']], on='Disaster_Type', how='left')
event_cols = ['Record_ID','DateKey','GeographyKey','DisasterTypeKey',
    'Compound_Event','Disaster_Duration_Days','Disaster_Magnitude','Hazard_Severity',
    'Number_of_Deaths','Number_of_Injuries','Number_of_People_Affected',
    'Displacement_Count','Houses_Damaged','Infrastructure_Damage_Score',
    'Crop_Loss_Percentage','Economic_Loss_Million']
fact_events = fact_events[[c for c in event_cols if c in fact_events.columns]]
fact_events.to_csv('data/FactDisasterEvents.csv', index=False)
print(f'OK: FactDisasterEvents: {len(fact_events)} rows')


OK: FactDisasterEvents: 2036 rows


In [41]:
# FactRiskPredictions (model forecasts & expected economic risk)
df_usable = engineer_features(df).dropna(subset=['Disaster_Next_Month']).copy()
X_all = df_usable[PRE_EVENT_PREDICTORS]

df_usable['Predicted_Disaster_Probability'] = best_pipe.predict_proba(X_all)[:, 1]
df_usable['Predicted_Disaster_Class'] = (df_usable['Predicted_Disaster_Probability'] >= opt_thresh).astype(int)
df_usable['Decision_Threshold'] = opt_thresh

loss_preds = best_reg.predict(X_all)
if getattr(best_reg, 'fit_log_', True):
    loss_preds = np.expm1(loss_preds)
df_usable['Predicted_Loss_Million'] = loss_preds
df_usable['Expected_Economic_Risk_Million'] = df_usable['Predicted_Disaster_Probability'] * df_usable['Predicted_Loss_Million']

def get_pred_type(year):
    if year <= 2022: return 'Training fitted'
    elif year == 2023: return 'Calibration'
    elif year == 2024: return 'Validation'
    else: return 'Final test'

df_usable['Prediction_Type'] = df_usable['Year'].apply(get_pred_type)
df_usable['Is_Out_Of_Sample'] = (df_usable['Year'] > 2022).astype(int)
df_usable['Model_Version'] = '3.1'
df_usable['Training_End_Date'] = '2022-12-01'

def get_alert_level(prob):
    if prob < opt_thresh: return 'Low'
    elif prob < opt_thresh + 0.15: return 'Moderate'
    elif prob < opt_thresh + 0.30: return 'High'
    else: return 'Critical'

df_usable['Alert_Level'] = df_usable['Predicted_Disaster_Probability'].apply(get_alert_level)
df_usable = df_usable.merge(dim_geo[['District','GeographyKey']], on='District', how='left')
df_usable['Prediction_DateKey'] = pd.to_datetime(df_usable['Event_Date']).dt.strftime('%Y%m%d').astype(int)

forecast_dates = pd.to_datetime(df_usable['Event_Date']) + pd.DateOffset(months=1)
df_usable['Forecast_DateKey'] = forecast_dates.dt.strftime('%Y%m%d').astype(int)

fact_pred = df_usable[[
    'Record_ID', 'GeographyKey', 'Prediction_DateKey', 'Forecast_DateKey',
    'Predicted_Disaster_Probability', 'Predicted_Disaster_Class', 'Alert_Level',
    'Decision_Threshold', 'Disaster_Next_Month', 'Expected_Economic_Risk_Million',
    'Prediction_Type', 'Is_Out_Of_Sample', 'Model_Version', 'Training_End_Date'
]].copy()
fact_pred.columns = [
    'Prediction_ID', 'GeographyKey', 'Prediction_DateKey', 'Forecast_DateKey',
    'Predicted_Disaster_Probability', 'Predicted_Disaster_Class', 'Alert_Level',
    'Decision_Threshold', 'Actual_Disaster_Occurred', 'Expected_Economic_Risk_Million',
    'Prediction_Type', 'Is_Out_Of_Sample', 'Model_Version', 'Training_End_Date'
]
fact_pred.to_csv('data/FactRiskPredictions.csv', index=False)
print(f'OK: FactRiskPredictions: {len(fact_pred)} rows')
print('\n=== Power BI Export Complete ===')
print('Star schema tables: DimDate, DimGeography, DimDisasterType, FactDistrictMonthRisk, FactDisasterEvents, FactRiskPredictions')


OK: FactRiskPredictions: 13100 rows

=== Power BI Export Complete ===
Star schema tables: DimDate, DimGeography, DimDisasterType, FactDistrictMonthRisk, FactDisasterEvents, FactRiskPredictions


**MSc Statistics Student Interpretation (Power BI Schema):**
We export the data in a clean star-schema design to support BI analytics:
- `DimGeography` contains the static district attributes (including the newly generated $k=4$ cluster labels).
- `FactDistrictMonthRisk` contains the monthly risk scores and lead predictions (13,200 rows).
- `FactDisasterEvents` contains post-event impact metrics (event-months only, $N=2764$).
This dimensional model prevents database redundancy and ensures correct SQL joins (preventing double-counting errors during aggregations).
